code/modules/modular_computers/file_system/programs/research_app.dm

```diff
@@ -0,0 +1,218 @@
+
+/datum/computer_file/program/science/ui_act(action, list/params)
+	. = ..()
+	if (.)
+		return
+	var/obj/item/computer_hardware/card_slot/card_slot
+	if(computer)
+		card_slot = computer.all_components[MC_CARD]
+	var/obj/item/card/id/user_id_card = card_slot.stored_card
+
+	// Check if the console is locked to block any actions occuring
+	if (locked && action != "toggleLock")
+		computer.say("Console is locked, cannot perform further actions.")
+		return TRUE
+
+	switch (action)
+		if ("toggleLock")
+			if(computer.obj_flags & EMAGGED)
+				to_chat(usr, "<span class='boldwarning'>Security protocol error: Unable to access locking protocols.</span>")
+				return TRUE
+			if(ACCESS_RND in user_id_card.access)
+				locked = !locked
+			else
+				to_chat(usr, "<span class='boldwarning'>Unauthorized Access. Please insert research ID card.</span>")
+			return TRUE
+		if ("researchNode")
+			if(!SSresearch.science_tech.available_nodes[params["node_id"]])
+				return TRUE
+			research_node(params["node_id"], usr)
+			return TRUE
+
+/datum/computer_file/program/science/ui_static_data(mob/user)
+	. = list(
+		"static_data" = list()
+	)
+
+	// Build node cache...
+	// Note this looks a bit ugly but its to reduce the size of the JSON payload
+	// by the greatest amount that we can, as larger JSON payloads result in
+	// hanging when the user opens the UI
+	var/node_cache = list()
+	for (var/nid in SSresearch.techweb_nodes)
+		var/datum/techweb_node/n = SSresearch.techweb_nodes[nid] || SSresearch.error_node
+		var/cid = "[compress_id(n.id)]"
+		node_cache[cid] = list(
+			"name" = n.display_name,
+			"description" = n.description
+		)
+		if (n.research_costs?.len)
+			node_cache[cid]["costs"] = list()
+			for (var/c in n.research_costs)
+				node_cache[cid]["costs"]["[compress_id(c)]"] = n.research_costs[c]
+		if (n.prereq_ids?.len)
+			node_cache[cid]["prereq_ids"] = list()
+			for (var/pn in n.prereq_ids)
+				node_cache[cid]["prereq_ids"] += compress_id(pn)
+		if (n.design_ids?.len)
+			node_cache[cid]["design_ids"] = list()
+			for (var/d in n.design_ids)
+				node_cache[cid]["design_ids"] += compress_id(d)
+		if (n.unlock_ids?.len)
+			node_cache[cid]["unlock_ids"] = list()
+			for (var/un in n.unlock_ids)
+				node_cache[cid]["unlock_ids"] += compress_id(un)
+		if (n.required_experiments?.len)
+			node_cache[cid]["required_experiments"] = n.required_experiments
+		if (n.discount_experiments?.len)
+			node_cache[cid]["discount_experiments"] = n.discount_experiments
+
+	// Build design cache
+	var/design_cache = list()
+	var/datum/asset/spritesheet/research_designs/ss = get_asset_datum(/datum/asset/spritesheet/research_designs)
+	var/size32x32 = "[ss.name]32x32"
+	for (var/did in SSresearch.techweb_designs)
+		var/datum/design/d = SSresearch.techweb_designs[did] || SSresearch.error_design
+		var/cid = "[compress_id(d.id)]"
+		var/size = ss.icon_size_id(d.id)
+		design_cache[cid] = list(
+			d.name,
+			"[size == size32x32 ? "" : "[size] "][d.id]"
```

Comment:
```suggestion
	var/datum/asset/spritesheet/research_designs/spritesheet= get_asset_datum(/datum/asset/spritesheet/research_designs)
	var/size32x32 = "[spritesheet.name]32x32"
	for (var/did in SSresearch.techweb_designs)
		var/datum/design/design = SSresearch.techweb_designs[did] || SSresearch.error_design
		var/cid = "[compress_id(design.id)]"
		var/size = spritesheet.icon_size_id(design.id)
		design_cache[cid] = list(
			design.name,
			"[size == size32x32 ? "" : "[size] "][design.id]"
```
This line does not meet our contribution standards, specifically for [variable name compliance](https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names), and should be refactored.